encoding/ccf/encode.go

```diff
@@ -0,0 +1,1766 @@
+//	fields: [
+//	    + [
+//	        name: tstr,
+//	        type: type-value
+//	    ]
+//	]
+func (e *Encoder) encodeFieldTypeValues(fieldTypes []cadence.Field, visited ccfTypeIDByCadenceType) error {
+	// Encode CBOR array head with number of field types.
+	err := e.enc.EncodeArrayHead(uint64(len(fieldTypes)))
+	if err != nil {
+		return err
+	}
+
+	if len(fieldTypes) == 1 {
+		// Avoid overhead of sorting if there is only one field type
+		return e.encodeFieldTypeValue(fieldTypes[0], visited)
+	}
+
+	// "Deterministic CCF Encoding Requirements" in CCF specs:
+	//
+	//   "composite-type-value.fields MUST be sorted by name."
+
+	// NOTE: bytewiseFieldIdentifierSorter doesn't sort fieldTypes in place.
+	// bytewiseFieldIdentifierSorter.indexes is used as sorted fieldTypes
+	// index.
+	sorter := newBytewiseFieldSorter(fieldTypes)
+
+	sort.Sort(sorter)
+
+	// Encode sorted field types.
+	for i := 0; i < len(sorter.indexes); i++ {
+		index := sorter.indexes[i]
+
+		err = e.encodeFieldTypeValue(fieldTypes[index], visited)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// encodeFieldTypeValue encodes one composite field type as
+// language=CDDL
+//
+//	[
+//	    name: tstr,
+//	    type: type-value
+//	]
+func (e *Encoder) encodeFieldTypeValue(fieldType cadence.Field, visited ccfTypeIDByCadenceType) error {
+	// Encode CBOR array head with length 2.
+	err := e.enc.EncodeRawBytes([]byte{
+		// array, 2 items follow
+		0x82,
+	})
+	if err != nil {
+		return err
+	}
+
+	// element 0: field name as tstr.
+	err = e.enc.EncodeString(fieldType.Identifier)
+	if err != nil {
+		return err
+	}
+
+	// element 1: field type as type-value.
+	return e.encodeTypeValue(fieldType.Type, visited)
+}
+
+// encodeInitializerTypeValues encodes composite initializers as
+// language=CDDL
+//
+//	initializers: [
+//	    * [
+//	        * [
+//	            label: tstr,
+//	            identifier: tstr,
+//	            type: type-value
+//	        ]
+//	    ]
+//	]
+func (e *Encoder) encodeInitializerTypeValues(initializerTypes [][]cadence.Parameter, visited ccfTypeIDByCadenceType) error {
+	// Encode CBOR array head with number of initializers.
+	err := e.enc.EncodeArrayHead(uint64(len(initializerTypes)))
+	if err != nil {
+		return err
+	}
+
+	// Encode initializers.
+	for i := 0; i < len(initializerTypes); i++ {
+		params := initializerTypes[i]
+
+		err = e.encodeParameterTypeValues(params, visited)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// encodeParameterTypeValues encodes composite initializer parameter types as
+// language=CDDL
+//
+//	 [
+//	    * [
+//	        label: tstr,
+//	        identifier: tstr,
+//	        type: type-value
+//	    ]
+//	]
+func (e *Encoder) encodeParameterTypeValues(parameterTypes []cadence.Parameter, visited ccfTypeIDByCadenceType) error {
+	// Encode CBOR array head with number of parameter types.
+	err := e.enc.EncodeArrayHead(uint64(len(parameterTypes)))
+	if err != nil {
+		return err
+	}
+
```

Comment:
Maybe short-circuit for `len(parameterTypes) == 0`, i.e. just return (and avoid allocation for sorting)

Reply:
Done in commit 266fb6a280996e376da1e56ac228bd006ee5c2b0.